cpp/src/io/parquet/page_enc.cu

```diff
@@ -2141,6 +2164,304 @@ CUDF_KERNEL void __launch_bounds__(block_size, 8)
+    s->page_start_val  = row_to_value_idx(s->page.start_row, s->col);
+    s->chunk_start_val = row_to_value_idx(s->ck.start_row, s->col);
+
+    // set pointer to beginning of scratch space (aligned to size_type boundary)
+    auto scratch_start =
+      reinterpret_cast<uintptr_t>(s->page.page_data + s->page.max_hdr_size + s->page.max_data_size);
+    scratch_start = util::round_up_unsafe(scratch_start, sizeof(size_type));
+    scratch_data  = reinterpret_cast<uint8_t*>(scratch_start);
+  }
+  __syncthreads();
+
+  // create offsets map (if needed)
+  // We only encode valid values, and we need to know adjacent valid strings. So first we'll
+  // create a mapping of leaf indexes to valid indexes:
+  //
+  // validity array is_valid:
+  //   1 1 0 1 0 1 1 0
+  //
+  // exclusive scan on is_valid yields mapping of leaf index -> valid index:
+  //   0 1 2 2 3 3 4 5
+  //
+  // Last value should equal page.num_valid. Now we need to transform that into a reverse
+  // lookup that maps valid index -> leaf index (of length num_valid):
+  //   0 1 3 5 6
+  //
+  auto const has_leaf_nulls = s->page.num_valid != s->page.num_leaf_values;
+
+  size_type* const offsets_map =
+    has_leaf_nulls ? reinterpret_cast<size_type*>(scratch_data) : nullptr;
+
+  if (offsets_map != nullptr) {
+    size_type* const forward_map = offsets_map + s->page.num_valid;
+
+    // create the validity array
+    for (int idx = t; idx < s->page.num_leaf_values; idx += block_size) {
+      size_type const idx_in_col = s->page_start_val + idx;
+      bool const is_valid =
+        idx_in_col < s->col.leaf_column->size() and s->col.leaf_column->is_valid(idx_in_col);
+      forward_map[idx] = is_valid ? 1 : 0;
+    }
+    __syncthreads();
+
+    // exclusive scan to get leaf_idx -> valid_idx
+    block_excl_sum<block_size>(forward_map, s->page.num_leaf_values, 0);
+
+    // now reverse map to get valid_idx -> leaf_idx mapping
+    for (int idx = t; idx < s->page.num_leaf_values; idx += block_size) {
+      size_type const idx_in_col = s->page_start_val + idx;
+      bool const is_valid =
+        idx_in_col < s->col.leaf_column->size() and s->col.leaf_column->is_valid(idx_in_col);
+      if (is_valid) { offsets_map[forward_map[idx]] = idx; }
+    }
+    __syncthreads();
+  }
+
+  size_type* const prefix_lengths =
+    has_leaf_nulls ? offsets_map + s->page.num_valid : reinterpret_cast<size_type*>(scratch_data);
+
+  auto const type_id = s->col.leaf_column->type().id();
+
+  auto const get_byte_array = [type_id, s](size_type idx) -> byte_array {
+    if (type_id == type_id::STRING) {
+      auto const str = s->col.leaf_column->element<string_view>(idx);
+      return {reinterpret_cast<uint8_t const*>(str.data()), str.size_bytes()};
+    } else if (s->col.output_as_byte_array && type_id == type_id::LIST) {
+      auto const str = get_element<statistics::byte_array_view>(*s->col.leaf_column, idx);
+      return {reinterpret_cast<uint8_t const*>(str.data()),
```

Comment:
Parquet uses unsigned bytes for BYTE_ARRAY and FIXED_LEN_BYTE_ARRAY. We're doing `==` rather than `compare`, but I'd prefer to keep the byte array representation in line with what a byte array in Parquet is.